tests/Doctrine/Migrations/Tests/DependencyFactoryTest.php

```diff
@@ -125,6 +127,46 @@ public function testEntityManager() : void
         self::assertTrue($di->isFrozen());
     }
 
+    public function testNoSchemaProviderRaiseException() : void
+    {
+        $di = DependencyFactory::fromConnection(new ExistingConfiguration($this->configuration), new ExistingConnection($this->connection));
+        self::assertFalse($di->hasSchemaProvider());
+
+        $this->expectException(MissingDependency::class);
+        $di->getSchemaProvider();
+    }
+
+    public function testSchemaProviderFromEntityManager() : void
+    {
+        $di = DependencyFactory::fromEntityManager(new ExistingConfiguration($this->configuration), new ExistingEntityManager($this->entityManager));
+
+        self::assertTrue($di->hasSchemaProvider());
+        self::assertEquals(new OrmSchemaProvider($this->entityManager), $di->getSchemaProvider());
+        self::assertTrue($di->isFrozen());
+    }
+
+    public function testSchemaProviderFromConfig() : void
+    {
+        $schemaProvider = $this->createMock(SchemaProvider::class);
+        $this->configuration->setSchemaProvider($schemaProvider);
+        $di = DependencyFactory::fromConnection(new ExistingConfiguration($this->configuration), new ExistingConnection($this->connection));
+
+        self::assertTrue($di->hasSchemaProvider());
+        self::assertSame($schemaProvider, $di->getSchemaProvider());
+        self::assertTrue($di->isFrozen());
+    }
+
+    public function testSchemaProviderWithEntityManagerConfigShouldFallBackToTheExplicitlyProvidedProvider() : void
```

Comment:
That's not really a fallback mechanism if the schema provider from the entity manager exists.

```suggestion
    public function testProvidedProviderHasPrecedenceOverProviderFromEntityManagerConfig() : void
```